src/loader/tsbgw_message_queue.c

```diff
@@ -0,0 +1,273 @@
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <storage/lwlock.h>
+#include <storage/shmem.h>
+#include <storage/proc.h>
+#include <storage/procarray.h>
+#include <storage/shm_mq.h>
+#include <access/xact.h>
+
+#include "../compat.h"
+
+#include "tsbgw_message_queue.h"
+
+#define TSBGW_MAX_MESSAGES 16
+#define TSBGW_MESSAGE_QUEUE_NAME "timescale_bgw_message_queue"
+#define TSBGW_MQ_TRANCHE_NAME "timescale_bgw_mq_tranche"
+
+#define TSBGW_ACK_QUEUE_SIZE (MAXALIGN(shm_mq_minimum_size + sizeof(int)))
+typedef struct TsbgwMessageQueue
+{
+	pid_t		reader_pid;		/* should only be set once at cluster launcher
+								 * startup */
+	LWLock	   *lock;			/* pointer to the lock to control
+								 * adding/removing elements from queue */
+	uint8		read_upto;
+	uint8		num_elements;
+	TsbgwMessage buffer[TSBGW_MAX_MESSAGES];
+}			TsbgwMessageQueue;
+
+typedef enum QueueResponseType
+{
+	MESSAGE_SENT = 0,
+	QUEUE_FULL,
+	READER_DETACHED
+}			QueueResponseType;
+
+static TsbgwMessageQueue * tsbgw_mq = NULL;
+
+static void
+queue_init(bool reinit)
```

Comment:
I still don't understand this `reinit` boolean.

Is there ever a case for calling this function were you are not initializing anything?

I see two cases where this function is used:

1. On startup, init is called with `reinit=false`, but you are still initializing because found=false.
2. In other cases, `reinit=true`, so you are also initializing.

In both case you are initializing, so why the boolean? Why not just do `queue_init()` to mean "initialize" no matter what? This would be much more clear. It seems neither `found` nor `reinit` really matters in any case.

Reply:
This is another thing I haven't yet addressed. I can do that shortly.